scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryClient.scala

```diff
@@ -456,14 +446,23 @@ class BigQueryClient private (private val projectId: String,
     }
   }
 
-  /** Extract tables touched by a query. */
+  /** Extract tables to be accessed by a query. */
   def extractTables(sqlQuery: String): Set[TableReference] = {
-    if (isLegacySql(sqlQuery, flattenResults = false)) {
-      BigQueryUtil.extractLegacyTables(sqlQuery)
-    } else {
-      val job = dryRunCache.get((sqlQuery, false, false)).get
-      job.getStatistics.getQuery.getReferencedTables.asScala.toSet
-    }
+    val isLegacy = isLegacySql(sqlQuery, flattenResults = false)
+    val tryJob = runQuery(sqlQuery, null, flattenResults = false, isLegacy, dryRun = true)
+    tryJob.get.getStatistics.getQuery.getReferencedTables.asScala.toSet
+  }
+
+  /** Extract locations of tables to be access by a query. */
+  def extractLocation(sqlQuery: String): String = {
+    val locations = extractTables(sqlQuery)
+      .map(t => (t.getProjectId, t.getDatasetId))
+      .map { case (pId, dId) =>
+        val l = bigquery.datasets().get(pId, dId).execute().getLocation
+        if (l != null) l else "US"
+      }
+    require(locations.size == 1, "Tables in the query must be in the same location")
```

Comment:
Could this be more human friendly error message - maybe list tables and locations?